# RAB TECH Task 02 – Data Quality Profile

## Objective
Profile the raw retail orders dataset to identify data quality issues
across completeness, uniqueness, validity, consistency, and freshness.

## Dataset
RAB TECH – Retail Orders Raw Dataset

In [27]:
import pandas as pd
df = pd.read_csv('retail-orders-raw.csv')
df.head()

,order_id,order_date,customer_segment,city,category,quantity,unit_price,discount_pct,payment_status
0,RT-1001,03-01-2026,Student,Chennai,Learning Kit,2,799,10.0,Paid
1,RT-1002,03-01-2026,Fresher,Bengaluru,Course Access,1,1499,0.0,paid
2,RT-1003,05-01-2026,student,Chennai,Course Access,1,1499,NaN,Pending
3,RT-1004,07-01-2026,Professional,Hyderabad,Learning Kit,3,799,5.0,Paid
4,RT-1004,07-01-2026,Professional,Hyderabad,Learning Kit,3,799,5.0,Paid


## 1. Dataset Overview

In [23]:
df.shape

(12, 9)

In [24]:
df.columns

Index(['order_id', 'order_date', 'customer_segment', 'city', 'category',
       'quantity', 'unit_price', 'discount_pct', 'payment_status'],
      dtype='object')

In [25]:
df.dtypes

,0
order_id,object
order_date,object
customer_segment,object
city,object
category,object
quantity,object
unit_price,int64
discount_pct,float64
payment_status,object


## 2. Completeness Check

In [2]:
df.isnull().sum()

,0
order_id,0
order_date,1
customer_segment,0
city,1
category,0
quantity,0
unit_price,0
discount_pct,1
payment_status,0


## 3. Uniqueness Check

In [3]:
df.duplicated().sum()

np.int64(1)

In [4]:
df.describe()

,unit_price,discount_pct
count,12.000000,11.000000
mean,1082.333333,13.636364
std,318.614425,30.748245
min,799.000000,0.000000
25%,799.000000,0.000000
50%,999.000000,5.000000
75%,1499.000000,10.000000
max,1499.000000,105.000000


## 4. Validity Checks

In [5]:
df["quantity"].min()

'-1'

In [6]:
df["unit_price"].min()

799

In [7]:
df["payment_status"].value_counts()

,count
payment_status,
Paid,7
Pending,2
paid,1
Failed,1
Refunded,1


In [8]:
df["customer_segment"].value_counts()

,count
customer_segment,
Student,4
Professional,4
Fresher,3
student,1


In [9]:
df["category"].value_counts()

,count
category,
Learning Kit,5
Course Access,4
Mentor Session,3


In [10]:
df["order_date"].dtype

dtype('O')

In [11]:
pd.to_datetime(df["order_date"], errors="coerce").isnull().sum()

np.int64(5)

In [12]:
df["discount_pct"].min(), df["discount_pct"].max()

(0.0, 105.0)

In [13]:
df[["quantity", "unit_price", "discount_pct"]].corr()

ValueError: could not convert string to float: 'two'

In [14]:
for col in ["quantity", "unit_price", "discount_pct"]:
    print(col, df[col].astype(str).str.lower().eq("two").sum())

quantity 1
unit_price 0
discount_pct 0


In [15]:
df["order_id"].duplicated().sum()

np.int64(1)

In [16]:
df[df["order_id"].duplicated(keep=False)]

,order_id,order_date,customer_segment,city,category,quantity,unit_price,discount_pct,payment_status
3,RT-1004,07-01-2026,Professional,Hyderabad,Learning Kit,3,799,5.0,Paid
4,RT-1004,07-01-2026,Professional,Hyderabad,Learning Kit,3,799,5.0,Paid


In [17]:
df[["quantity", "unit_price", "discount_pct"]]

,quantity,unit_price,discount_pct
0,2,799,10.0
1,1,1499,0.0
2,1,1499,NaN
3,3,799,5.0
4,3,799,5.0
5,1,999,0.0
6,-1,799,10.0
7,2,999,105.0
8,two,1499,0.0
9,1,799,0.0


In [18]:
df["payment_status"].value_counts(normalize=True) * 100

,proportion
payment_status,
Paid,58.333333
Pending,16.666667
paid,8.333333
Failed,8.333333
Refunded,8.333333


In [19]:
df["city"].value_counts(dropna=False)

,count
city,
Chennai,3
Bengaluru,2
Hyderabad,2
NaN,1
Pune,1
Mumbai,1
Delhi,1
Kochi,1


In [20]:
df["order_date"].value_counts(dropna=False)

,count
order_date,
03-01-2026,2
07-01-2026,2
05-01-2026,1
09-01-2026,1
2026-13-10,1
12-01-2026,1
14-01-2026,1
16-01-2026,1
18-01-2026,1


In [21]:
df["order_date"].min(), df["order_date"].max()

TypeError: '<=' not supported between instances of 'str' and 'float'

## 5. Freshness Check

In [26]:
converted_dates = pd.to_datetime(
    df["order_date"],
    format="%d-%m-%Y",
    errors="coerce"
)

converted_dates.min(), converted_dates.max()

(Timestamp('2026-01-03 00:00:00'), Timestamp('2026-01-18 00:00:00'))

## 6. Data Quality Findings Summary

The profiling identified the following data quality issues:

1. Missing values in order_date, city, and discount_pct.
2. One duplicate order record and one duplicated order_id (RT-1004).
3. Invalid quantity values: -1 and "two".
4. Invalid discount value: 105%.
5. Inconsistent categorical values: "paid" and "student".
6. One invalid order date: 2026-13-10.
7. order_date is stored as an object/string and requires date standardization.
8. Valid order dates range from 03-Jan-2026 to 18-Jan-2026.

No data was modified during profiling. Data cleaning and correction will be performed in Task 03.